crate2nix/templates/nix/crate2nix/default.nix

```diff
@@ -70,65 +69,88 @@ rec {
     );
 
   /* A restricted overridable version of  buildRustCrateWithFeaturesImpl. */
-  buildRustCrateWithFeatures = {
-        packageId, 
-        features ? rootFeatures,
-        crateOverrides ? defaultCrateOverrides, 
-        buildRustCrateFunc ? buildRustCrate
-      }:
+  buildRustCrateWithFeatures =
+    { packageId
+    , features ? rootFeatures
+    , crateOverrides ? defaultCrateOverrides
+    , buildRustCrateFunc ? buildRustCrate
+    , doTest ? false,
+    }:
     lib.makeOverridable
-      ({features, crateOverrides}: 
+      ({features, crateOverrides, doTest}:
         let builtRustCrates = builtRustCratesWithFeatures {
-          inherit packageId features crateOverrides  buildRustCrateFunc;
+          inherit packageId features crateOverrides buildRustCrateFunc doTest;
         };
         in builtRustCrates.${packageId})
-      { inherit features crateOverrides; };
+      { inherit features crateOverrides doTest; };
 
   /* Returns a buildRustCrate derivation for the given packageId and features. */
-  builtRustCratesWithFeatures = { 
-        crateConfigs? crates, 
-        packageId,
-        features,
-        crateOverrides, 
-        buildRustCrateFunc
-      } @ args:
+  builtRustCratesWithFeatures =
+    { packageId
+    , features
+    , crateConfigs ? crates
+    , crateOverrides
+    , buildRustCrateFunc
+    , doTest
+    , target ? defaultTarget
+    } @ args:
     assert (builtins.isAttrs crateConfigs);
     assert (builtins.isString packageId);
     assert (builtins.isList features);
+    assert (builtins.isAttrs target);
+
+    let mergedFeatures = mergePackageFeatures ( args // { inherit target; });
+        mergedTestFeatures = mergePackageFeatures ( args // { target = target // { test = true; }; });
+
+        buildByPackageId = packageId: buildByPackageIdImpl packageId false;
+        buildTestByPackageId = packageId: buildByPackageIdImpl packageId true;
```

Comment:
This is still unused, correct?

Reply:
> This is super cool, thank you.

Thank you for all the work that I didn't have to do :heart: 

> For merging (now), I would, for now, either make `doTest` do something or remove it. (Or did I miss that it already does something? ) Then we can already merge and you'll have less potential merge conflicts down the way.

I would probably postpone merging just a few more days until I've the nixpkgs change drafted. Then I have all the pieces to present and we can probably get a much better initial interface and do not have to worry about backwards compatibility (for now).

> I guess that if "doTest" is true, we should optimally create a dervivation that depends both on the release build and the tests -- that way they can build in parallel:
> 
>     * crate derivation, outputs include release binaries and maybe as a different output (never used multiple outputs) the test results
>       
>       * release build derivation with output binaries (what we have currently)
>         
>         * release dependencies ....
>       * run tests derivation that runs the tests, output are the test results
>         
>         * a derivation for each test?
>           
>           * test dependencies ...

My plan was to inject a version of `rootCrate` that has `target.test=true` enabled into `buildRustCrate`. The contract here would be that additionally passed derivation resembles a configuration of the crate that has test support enabled.
`buildRustCrate` could then either override that derivation to fit its needs (extraRustcArgs = [ "--test" ];, modified install phase, added check phase that runs all the tests). On the other hand we could also implement most of that in `crate2nix`. I am currently looking more towards having it in nixpkgs as that allows us to have a common interface among many rust2nix tools.


> Given how buildRustCrate is currently structured (information is passed from build.rs to the rest of the build via env variables), this might be difficult to achieve. For example, I think one would need to capture the results of build.rs so that they are available in other derivations without rerunning build.rs.

I did that as an experiment some time ago and that wasn't too hard. I basically added a few arguments to `build.rs` that allowed running tests instead of producing outputs.

Because that was rather hacky I started reworking the entire `buildRustCrate` expression. The exact interface I am aiming for isn't set yet. I believe the implemention can be enhanced incrementally as long as we have a proper interface.

> 
> Thus, we should probably aim lower and just execute tests in the same derivation that we build the test binaries? Anyway, if there is potential for easy splitting into multiple derivations there, I am all for it.

As it stands every rust crate can have an entirely different set of dependencies during testing. I would try to avoid mixing both of them and thus in favor of having tests run in a dedicated expression.